### PR TITLE
Heavy path decomposition

### PR DESCRIPTION
**Pull request by AlexandruValeanu**
I've finished the heavy path decomposition implementation. This will be used to implement a very efficient LCA finder (which will be part a future PR).  

There are a lot of getters so please let me know if you have any suggestions about them or their docs.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    }
+
+    /**
+     * Returns the father of vertex $v$ in the internal DFS tree/forest.
+     * If the vertex $v$ has not been explored or it is the root of its tree, $null$ will be returned.
+     *
+     * @param v vertex
+     * @return father of vertex $v$ in the DFS tree/forest
+     */
+    public V getFather(V v){
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Let's make our trees gender-neutral with parent/child instead.

**Comment by d-michail** [comment]
@jsichi @jkinable Would it make sense to split this into two classes, one for the algorithm and one for the result (similar to what we are doing with our interfaces)? Otherwise the user runs the algorithm and then has to store 5-6 different arrays, perhaps copy them, etc.

**Comment by jsichi** [comment]
Makes sense.  Seems likely that we'll want one or more interfaces here eventually.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+                    if (!explored.contains(child)){
+                        parent[child] = u;
+                        depth[child] = depth[u] + 1;
+                        stack.push(child);
+                    }
+                }
+            }
+            else{
+                /*
+                    For u compute heavySon. If it exists then u becomes part of heavySon's path.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Let's go with heavyChild for consistency.

**Comment by d-michail** [comment]
This PR raises the interesting question on how to provide a high-level enough interface for the result while at the same time retaining the low-level representation in order to be useful in other algorithms. 

 - As a user I would expect to call the algorithm and get back a list of heavy edges (Set<E>). I would probably create an interface called `HeavyEdges` just like our other interfaces.
 - On the other hand, you as a developer of an algorithm prefer to have an indexed representation (into multiple arrays), in order to reuse the representation. 

How are we going to handle this?  I see three alternatives: 
 - Keep the low-level representation as it is. I am not really fond of this.
 - Return the result using an interface just like `VertexCover` or `Matching` and get a set of heavy edges.
 - Return the result using an interface which is a set of heavy edges, but at the same time also provides access to the low-level representation.

@AlexandruValeanu @jsichi @jkinable Ideas? Preferences?

**Comment by jsichi** [comment]
For results which are expected to be used from outside JGraphT (or from other packages within JGraphT), the public methods should not return low-level representations.  For results which are used only from within the same package, it's OK to expose low-level representations, but only via package-private methods, so that we are free to evolve them agile-style without deprecation.

In this case, it sounds like the consumer will be in a different package (lca), so no low-level representation should be exposed.

For the moment, methods such as getParentArray should not even exist, since there's no caller (not even from a unit test).  So remove them for now, and then we can discuss the internal interface design (the "developer-oriented" result) as part of the PR that puts them into action for LCA finding.

Regarding the interface for the "user-oriented" result, were you thinking something like this?

* algorithm interface TreeToPathDecompositionAlgorithm (to distinguish from general graph-to-subgraph decompositions, including graph-to-tree decompositions)
* interface PathDecomposition for the result, which besides producing the set of edges (similar to Spanner), also produces the set-of-paths representation, and/or a "tree of paths" representation
* HeavyPathDecomposition implements TreeToPathDecompositionAlgorithm

At this level, we should avoid over-specific interfaces such as "HeavyEdges" since we want to make the result interface applicable to other kinds of decomposition as well.  But "HeavyEdges" might be appropriate for the "developer-oriented" result.

**Comment by AlexandruValeanu** [comment]
I have added the interface. I am not happy with the InternalState part as all types returned by any function inside that class are erased so so need explicit type conversions in the calling code.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 3]
```diff
@@ -64,6 +65,7 @@ public static int log2(int n) // returns 0 for n=0
         }
 
         int maxim = 0;
+        HeavyPathDecomposition.InternalState state = decomposition.getInternalState();
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: If you change this to HeavyPathDecomposition<V,E>.InternalState, won't that make your typecasting woes go away?

Comment (AlexandruValeanu) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I need to stop using Intellij's autocomplete blindly. Thank you, @jsichi!

Comment (jsichi) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I am glad we didn't have to resort to reflection :)

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+                }
+
+                maxim = Math.max(maxim, cnt);
+            }
+        }
+
+        return maxim;
+    }
+
+    public static <V, E> boolean isValidDecomposition(Graph<V, E> graph, Set<V> roots,
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I think you have to test for each node, that the size of the subtree rooted at a heavy edge is larger than the sum of sizes of all subtrees rooted at light edges. The log_2 property then follows.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        }
+
+        /**
+         * Return a copy of the internal depth array.
+         * For each vertex $v \in V$, $depthArray[normalizeVertex(v)] = getDepth(v)$
+         *
+         * @return internal depth array
+         */
+        public int[] getDepthArray(){
+            return depth.clone();
```
[the remark this answers is not in the document]
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: We do not expose private fields so we do not really break encapsulation.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+                    added (the order will be reversed in decompose).
+                 */
+                positionInPath[u] = lengthPath[path[u]]++;
+            }
+        }
+    }
+
+    private void decompose(){
+        /*
+            If we already have a decomposition stop.
```
[the remark this answers is not in the document]
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: I didn't forget it. It is supposed to be a regular comment, not a doc one.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java (hunk 4 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I don't think that works. Take for example [1, 2, 3, 4] {1 -> 2, 1->3, 1->4}. There is only one heavy edge (1 -> 2; any of them works) and the subtree rooted 2 doesn't contain more vertices than all other three combined. Let me know if I didn't understand what you meant.

The current test works as the number of heavy edges on any node-to-root path is at most log2(|V|.)

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: I think what @d-michail means is that since we're already providing access to internal state for the purpose of efficiency, might was well skip the clone and trust the caller not to modify the original.  I agree, let's skip the clone.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Docs are with two stars :). When you add a star in each line, the IDEs respect the comments during reformatting and render them much better.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+         * @return edge set of the path decomposition
+         */
+        Set<E> getEdges();
+
+        /**
+         * List of disjoint vertex paths of the decomposition
+         *
+         * @return list of vertex paths
+         */
+        List<List<V>> getPaths();
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: Is there a reason not to produce org.jgrapht.GraphPath<V,E> instead of List<V>?  org.jgrapht.graph.GraphWalk makes it easy to produce a GraphPath from a List<V>.

Also, technically it should be Set<GraphPath> rather than List<GraphPath>.  It's possible to do this efficiently with an ArrayUnenforcedSet (elements unique by construction) wrapped as unmodifiable.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: I get that. I am only saying that there is no difference between the comment that starts at line 230 and the one that starts at 260. There were not meant to be javadoc comments. When I add another * I get a dangling javadoc comment for example (Intellij Idea)

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java (hunk 4 above)
Comment (d-michail) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: You are right, my definition was wrong. A heavy edge is when the child has size larger than half of the size of the parent. Light edges are the remaining ones.

However, the number of light edges is log(n) on any root-to-leaf path. This is a consequence of the above definition, which means that if you follow a light edge you can have at most half the nodes.

So your test should be to verify that each light edge has this property. The logn bound follows from that. Just do a post-order and count the number of nodes of each subtree (including the subtree root) and then verify for each light edge.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: But add a comment in the Javadoc, to explicitly explain that.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java (hunk 7 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: Yes, I forgot about `GraphPath`. :)

I've made that change but I don't think that `PathDecompositionImpl` should be `Serializable` anymore since `GraphWalk` isn't. Any thoughts on this?

Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: Seems like GraphWalk should be made serializable as well.

Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: I can submit a PR for that.

Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: #628 solves this 'issue'.

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: `GraphWalk` contains a reference to the graph instance. Not sure it makes sense to make it serializable.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java (hunk 4 above)
Comment (d-michail) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I guess for the test to be complete, you also need to test the property of the heavy edges. Then the logn bound check will be superfluous.

File: jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java (hunk 7 above)
Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/TreeToPathDecompositionAlgorithm.java]: A number of our other algorithm return-values (cycle basis, spanner, spanning tree) are serializable with graph references.  I can see how in other contexts, a user might be unpleasantly surprised to find the entire graph serialized after they asked to only serialize a small walk by itself, but they'll get over it.  In the decomposition case, we're returning a bunch of paths all over the graph, and serialization takes care of only saving one copy of the graph automatically.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java (hunk 4 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: I'd like to keep that last test. If it is redundant at this stage.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+     */
+    @Override
+    public PathDecomposition<V, E> getPathDecomposition() {
+        return new PathDecompositionImpl<>(graph, getHeavyEdges(), this.paths);
+    }
+
+    /**
+     * Return the internal representation of the data.
+     *
+     * Note: For developer use only
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Add one more comment here which states that we are returning direct references to the internal data structures and that this is done only for efficiency reasons.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+
+                    if (!u.equals(bfs.getParent(v))){
+                        int sizeU = sizeSubtree.get(u);
+                        maxSizeSubtree = Math.max(maxSizeSubtree, sizeU);
+
+                        sizeSubtree.put(v, sizeSubtree.get(v) + sizeU);
+                    }
+                }
+
+                final int totalSize = sizeSubtree.get(v) - 1;
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: This is not needed.

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+
+                for (E edge: graph.edgesOf(v)){
+                    if (lightEdges.contains(edge)){
+                        V u = Graphs.getOppositeVertex(graph, edge, v);
+
+                        if (!u.equals(bfs.getParent(v)) && 2 * sizeSubtree.get(u) > totalSize) {
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: `&& 2 * sizeSubtree.get(u) <= sizeSubtree.get(v)`

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+                            return false;
+                        }
+                    }
+                    else{ // edge is heavy
+                        V u = Graphs.getOppositeVertex(graph, edge, v);
+
+                        if (!u.equals(bfs.getParent(v)) && sizeSubtree.get(u) < maxSizeSubtree) {
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: `&& 2 * sizeSubtree.get(u) > sizeSubtree.get(v)`

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+                V v = bfs.next();
+                postOrder.add(v);
+            }
+
+            Collections.reverse(postOrder);
+
+            Map<V, Integer> sizeSubtree = new HashMap<>(graph.vertexSet().size());
+            for (V v: postOrder){
+                sizeSubtree.put(v, 1);
+                int maxSizeSubtree = -1;
```
Comment (d-michail) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: Not needed at all.

**Comment by d-michail** [comment]
At https://courses.csail.mit.edu/6.851/spring12/scribe/L19.pdf you can find some notes from Demaine's advanced data structures course. In Section 4.1.1 you can find the definition of heavy-light edges. 

Otherwise the original paper by Sleator and Tarjan (https://www.cs.cmu.edu/~sleator/papers/dynamic-trees.pdf) contains the definition at page 370.

```
We define the size of a vertex v in the forest, denoted by size(v), to be the number of descendants
of v, including v itself. We define a tree edge (v,parent(v)) to be heavy if 2 size(v) > size(parent(v))
and light, otherwise.
```

Let us follow the definition and return what is actually a correct heavy-light edges decomposition.

Additionally, add one more test with this tree `[1, 2, 3, 4] {1 -> 2, 1 -> 3, 1 -> 4}` and make sure that the algorithm returns zero heavy edges, and 3 light edges.

**Comment by AlexandruValeanu** [comment]
@d-michail I have updated the definition of heavy-edges. I have also changed the tests a little bit and the class' description.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+        }
+
+        for (int i = 0; i < numberOfPaths; i++) {
+            paths.set(i, Collections.unmodifiableList(paths.get(i)));
+        }
+
+        this.paths = Collections.unmodifiableList(paths);
+
+        this.heavyEdges = Collections.unmodifiableSet(this.heavyEdges);
+        this.lightEdges = Collections.unmodifiableSet(graph.edgeSet().stream()
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Could we save this linear scan, by filling the lightEdges at the same time as the heavyEdges, perhaps with an else clause?

**Comment by d-michail** [comment]
Great! 

Could you also include in the class Javadoc the definition of heavy and light edges as written in the Sleator & Tarjan original paper?

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 13 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: Done

**Comment by AlexandruValeanu** [comment]
@d-michail I have updated the docs.

**Comment by d-michail** [comment]
@jsichi Any additional comments?

**Comment by AlexandruValeanu** [comment]
@jsichi Thank you for your comments. I believe I've addressed all of them.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 14]
```diff
@@ -0,0 +1,538 @@
...
+        path = new int[n];
+        lengthPath = new int[n];
+        positionInPath = new int[n];
+
+        heavyEdges = new HashSet<>();
+        lightEdges = new HashSet<>();
+    }
+
+    private void normalizeGraph(){
+        /**
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: There is some kind of communication barrier here :)  This comment was previously in the desired format:

        /*
         * Normalize the graph: map each vertex to an integer (using a HashMap) keep the reverse
         * mapping (using an ArrayList)
         */

That's the way all of them should be if you are going with slash star format.

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java [hunk 15]
```diff
@@ -0,0 +1,538 @@
...
+
+                int pathChild = -1;
+                E pathEdge = null;
+
+                V vertexU = indexList.get(u);
+                for (E edge: graph.edgesOf(vertexU)){
+                    int child = vertexMap.get(Graphs.getOppositeVertex(graph, edge, vertexU));
+
+                    /**
+                        Check if child if a descendant of u and not its parent
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: "is"

File: jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java (hunk 14 above)
Comment (AlexandruValeanu) [reply jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java]: I finally understood what you meant. I decided to go with the classical //

File: jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Tests for {@link HeavyPathDecomposition}
+ *
+ * @author Alexandru Valeanu
+ */
+public class HeavyPathDecompositionTest {
+
+    public static int log2(int n) // returns 0 for n=0
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/decomposition/HeavyPathDecompositionTest.java]: We should probably move this to org.jgrapht.util.MathUtil

**Comment by jbduncan** [comment]
@jsichi 

> Maybe for next year's GSoC, someone can round up all the internal dfs code (which keeps proliferating) and see if it can be efficiently unified.

Perhaps the API for Guava's [`Traverser`](https://google.github.io/guava/releases/25.1-jre/api/docs/com/google/common/graph/Traverser.html) could prove to be a good starting point for this?

**Comment by jsichi** [comment]
@jbduncan well we already have traversal iterator support in JGraphT, but for use internally within algorithms, it's not always efficient, which is why people end up rolling their own using non-iterator patterns.

**Comment by jbduncan** [comment]
@jsichi Ah okay, I hadn't fully realised that JGraphT already had iterator support for traversals like DFS! Thanks for enlightening me on that. :)

**Comment by AlexandruValeanu** [comment]
@d-michail @jsichi Do you have any more comments on this PR?

**Comment by AlexandruValeanu** [comment]
@jsichi Sure. I have a couple of conflicts so I'll probably submit the new PR tomorrow evening.
